File: wasmtime-wasi/js-polyfill/wasi.js
// To implement `proc_exit`, we define a custom exception object
// that we can throw to unwind the stack and carry the exit value.
function WASIExit(return_value, message, fileName, lineNumber) {
    let instance = new Error(message, fileName, lineNumber);
    instance.return_value = return_value;
    Object.setPrototypeOf(instance, Object.getPrototypeOf(this));
    if (Error.captureStackTrace) {
        Error.captureStackTrace(instance, WASIExit);
    }
    return instance;
}

WASIExit.prototype = Object.create(Error.prototype, {
    constructor: {
      value: Error,
      enumerable: false,
      writable: true,
      configurable: true
    }
});

if (Object.setPrototypeOf) {
    Object.setPrototypeOf(WASIExit, Error);
} else {
    WASIExit.__proto__ = Error;
}

function handleWASIExit(e) {
    if (e.return_value != 0) {
        console.log('program exited with non-zero exit status ' + e.return_value);
    }
}

// The current guest wasm instance.
var currentInstance;

// There are two heaps in play, the guest heap, which belongs to the WASI-using
// program, and the host heap, which belongs to the Emscripten-compiled polyfill
// library. The following declare support for the guest heap in a similar manner
// to Emscripten's heap.

var GUEST_HEAP,
/** @type {ArrayBuffer} */
  GUEST_buffer,
/** @type {Int8Array} */
  GUEST_HEAP8,
/** @type {Uint8Array} */
  GUEST_HEAPU8,
/** @type {Int16Array} */
  GUEST_HEAP16,
/** @type {Uint16Array} */
  GUEST_HEAPU16,
/** @type {Int32Array} */
  GUEST_HEAP32,
/** @type {Uint32Array} */
  GUEST_HEAPU32,
/** @type {Float32Array} */
  GUEST_HEAPF32,
/** @type {Float64Array} */
  GUEST_HEAPF64;

function setInstance(instance) {
  currentInstance = instance;
  updateGuestBuffer();
}

/// We call updateGuestBuffer any time the guest's memory may have changed,
/// such as when creating a new instance, or after calling _malloc.
function updateGuestBuffer() {
  var buf = currentInstance.exports.memory.buffer;
  Module['GUEST_buffer'] = GUEST_buffer = buf;
  Module['GUEST_HEAP8'] = GUEST_HEAP8 = new Int8Array(GUEST_buffer);
  Module['GUEST_HEAP16'] = GUEST_HEAP16 = new Int16Array(GUEST_buffer);
  Module['GUEST_HEAP32'] = GUEST_HEAP32 = new Int32Array(GUEST_buffer);
  Module['GUEST_HEAPU8'] = GUEST_HEAPU8 = new Uint8Array(GUEST_buffer);
  Module['GUEST_HEAPU16'] = GUEST_HEAPU16 = new Uint16Array(GUEST_buffer);
  Module['GUEST_HEAPU32'] = GUEST_HEAPU32 = new Uint32Array(GUEST_buffer);
  Module['GUEST_HEAPF32'] = GUEST_HEAPF32 = new Float32Array(GUEST_buffer);
  Module['GUEST_HEAPF64'] = GUEST_HEAPF64 = new Float64Array(GUEST_buffer);
}

function copyin_bytes(src, len) {
    let dst = _malloc(len);
    updateGuestBuffer();

    for (let i = 0; i < len; ++i) {
        HEAP8[dst + i] = GUEST_HEAP8[src + i];
    }
    return dst;
}

function copyout_bytes(dst, src, len) {
    updateGuestBuffer();

    for (let i = 0; i < len; ++i) {
        GUEST_HEAP8[dst + i] = HEAP8[src + i];
    }
    _free(src);
}

function copyout_i32(dst, src) {
    updateGuestBuffer();

    GUEST_HEAP32[dst>>2] = HEAP32[src>>2];
    _free(src);
}

function copyout_i64(dst, src) {
    updateGuestBuffer();

    GUEST_HEAP32[dst>>2] = HEAP32[src>>2];
    GUEST_HEAP32[(dst + 4)>>2] = HEAP32[(src + 4)>>2];
    _free(src);
}

function translate_ciovs(iovs, iovs_len) {
    host_iovs = _malloc(8 * iovs_len);
    updateGuestBuffer();

    for (let i = 0; i < iovs_len; ++i) {
        let ptr = GUEST_HEAP32[(iovs + i * 8 + 0) >> 2];
        let len = GUEST_HEAP32[(iovs + i * 8 + 4) >> 2];
        let buf = copyin_bytes(ptr, len);
        HEAP32[(host_iovs + i * 8 + 0)>>2] = buf;
        HEAP32[(host_iovs + i * 8 + 4)>>2] = len;
    }
    return host_iovs;
}

function free_ciovs(host_iovs, iovs_len) {
    for (let i = 0; i < iovs_len; ++i) {
        let buf = HEAP32[(host_iovs + i * 8 + 0) >> 2];
        _free(buf);
    }
    _free(host_iovs);
}

function translate_iovs(iovs, iovs_len) {
    host_iovs = _malloc(8 * iovs_len);
    updateGuestBuffer();

    for (let i = 0; i < iovs_len; ++i) {
        let len = GUEST_HEAP32[(iovs + i * 8 + 4) >> 2];
        let buf = _malloc(len);
        updateGuestBuffer();
        HEAP32[(host_iovs + i * 8 + 0)>>2] = buf;
        HEAP32[(host_iovs + i * 8 + 4)>>2] = len;
    }
    return host_iovs;
}

function free_iovs(host_iovs, iovs_len, iovs) {
    updateGuestBuffer();
    for (let i = 0; i < iovs_len; ++i) {
        let buf = HEAP32[(host_iovs + i * 8 + 0) >> 2];
        let len = HEAP32[(host_iovs + i * 8 + 4) >> 2];
        let ptr = GUEST_HEAP32[(host_iovs + i * 8 + 0) >> 2];
        copyout_bytes(ptr, buf, len);
    }
    _free(host_iovs);
}

var WASIPolyfill = {

args_get: function(argv, argv_buf) {
    return 0;
},

args_sizes_get: function(argc, argv_buf_size) {
    updateGuestBuffer();

    // TODO: Implement command-line arguments.
    GUEST_HEAP32[(argc) >> 2] = 0;
    GUEST_HEAP32[(argv_buf_size) >> 2] = 0;
    return 0;
},

clock_res_get: function(clock_id, resolution) {
    let host_resolution = _malloc(8);
    let ret = ___wasi_clock_res_get(clock_id, host_resolution);
    copyout_i64(resolution, host_resolution);
    return ret;
},

clock_time_get: function(clock_id, precision, time) {
    let host_time = _malloc(8);
    let ret = ___wasi_clock_time_get(clock_id, precision, host_time);
    copyout_i64(time, host_time);
    return ret;
},

environ_get: function(environ, environ_buf) {
    return 0;
},

environ_sizes_get: function(environ_size, environ_buf_size) {
    updateGuestBuffer();

    // TODO: Implement environment variables.
    GUEST_HEAP32[(environ_size) >> 2] = 0;
    GUEST_HEAP32[(environ_buf_size) >> 2] = 0;
    return 0;
},

fd_prestat_get: function(fd, buf) {
    let host_buf = _malloc(8); // sizeof __wasi_prestat_t
    let ret = ___wasi_fd_prestat_get(fd, host_buf);
    copyout_bytes(buf, host_buf, 8);
    return ret;
},

fd_prestat_dir_name: function(fd, path, path_len) {
    let host_buf = _malloc(path_len);
    let ret = ___wasi_fd_prestat_get(fd, host_buf, path_len);
    copyout_bytes(buf, host_buf, path_len);
    return ret;
},

fd_close: function(fd) {
    return ___wasi_fd_close(fd);
},

fd_datasync: function(fd) {
    return ___wasi_fd_datasync(fd);
},

fd_pread: function(fd, iovs, iovs_len, offset, nread) {
    let host_iovs = translate_iovs(iovs, iovs_len);
    let host_nread = _malloc(4);
    let ret = ___wasi_fd_pread(fd, host_iovs, iovs_len, offset, host_nread);
    copyout_i32(nread, host_nread);
    free_iovs(host_iovs, iovs_len);
    return ret;
},

fd_pwrite: function(fd, iovs, iovs_len, offset, nwritten) {
    let host_iovs = translate_ciovs(iovs, iovs_len);
    let host_nwritten = _malloc(4);
    let ret = ___wasi_fd_pwrite(fd, host_iovs, iovs_len, offset, host_nwritten);
    copyout_i32(nwritten, host_nwritten);
    free_ciovs(host_iovs, iovs_len);
    return ret;
},

fd_read: function(fd, iovs, iovs_len, nread) {
    let host_iovs = translate_iovs(iovs, iovs_len);
    let host_nread = _malloc(4);
    let ret = ___wasi_fd_read(fd, host_iovs, iovs_len, host_nread);
    copyout_i32(nread, host_nread);
    free_iovs(host_iovs, iovs_len);
    return ret;
},

fd_renumber: function(from, to) {
    return ___wasi_fd_renumber(from, to);
},

fd_seek: function(fd, offset, whence, newoffset) {
    let host_newoffset = _malloc(8);
    let ret = ___wasi_fd_seek(fd, offset, whence, host_newoffset);
    copyout_i64(newoffset, host_newoffset);
    return ret;
},

fd_tell: function(fd, newoffset) {
    let host_newoffset = _malloc(8);
    let ret = ___wasi_fd_seek(fd, host_newoffset);
    copyout_i64(newoffset, host_newoffset);
    return ret;
},

fd_fdstat_get: function(fd, buf) {
    let host_buf = _malloc(24); // sizeof __wasi_fdstat_t
    let ret = ___wasi_fd_fdstat_get(fd, host_buf);
    copyout_bytes(buf, host_buf, 24);
    return ret;
},

fd_fdstat_set_flags: function(fd, flags) {
    return ___wasi_fd_fdstat_set_flags(fd, flags);
},

fd_fdstat_set_rights: function(fd, fs_rights_base, fs_rights_inheriting) {
    return ___wasi_fd_fdstat_set_rights(fd, fs_rights_base, fs_rights_inheriting);
},

fd_sync: function(fd) {
    return ___wasi_fd_sync(fd);
},

fd_write: function(fd, iovs, iovs_len, nwritten) {
    let host_iovs = translate_ciovs(iovs, iovs_len);
    let host_nwritten = _malloc(4);
    let ret = ___wasi_fd_write(fd, host_iovs, iovs_len, host_nwritten);
    copyout_i32(nwritten, host_nwritten);
    free_ciovs(host_iovs, iovs_len);
    return ret;
},

fd_advise: function(fd, offset, len, advice) {
    return ___wasi_fd_advise(fd, offset, len, advice);
},

fd_allocate: function(fd, offset, len) {
    return ___wasi_fd_allocate(fd, offset, len);
},

path_create_directory: function(fd, path, path_len) {
    let host_path = copyin_bytes(path, path_len);
    let ret = ___wasi_path_create_directory(fd, host_path, path_len);
    _free(host_path);
    return ret;
},

path_link: function(fd0, path0, path_len0, fd1, path1, path_len1) {
    let host_path0 = copyin_bytes(path0, path_len0);
    let host_path1 = copyin_bytes(path1, path_len1);
    let ret = ___wasi_path_link(fd, host_path0, path_len0, fd1, host_path1, path1_len);
    _free(host_path1);
    _free(host_path0);
    return ret;
},

path_open: function(dirfd, dirflags, path, path_len, oflags, fs_rights_base, fs_rights_inheriting, fs_flags, fd) {
    let host_path = copyin_bytes(path, path_len);
    let host_fd = _malloc(4);
    let ret = ___wasi_path_open(dirfd, dirflags, host_path, path_len, oflags, fs_rights_base, fs_rights_inheriting, fs_flags, host_fd);
    copyout_i32(fd, host_fd);
    _free(host_path);
    return ret;
},

fd_readdir: function(fd, buf, buf_len, cookie, buf_used) {
    let host_buf = _malloc(buf_len);
    let host_buf_used = _malloc(4);
    let ret = ___wasi_fd_readdir(fd, buf, buf_len, cookie, host_buf_used);
    copyout_i32(buf_used, host_buf_used);
    copyout_bytes(buf, host_buf, buf_len);
    return ret;
},

path_readlink: function(fd, path, path_len, buf, buf_len, buf_used) {
    let host_path = copyin_bytes(path, path_len);
    let host_buf = _malloc(buf_len);
    let host_buf_used = _malloc(4);
    let ret = ___wasi_path_readlink(fd, path, path_len, buf, buf_len, host_buf_used);
    copyout_i32(buf_used, host_buf_used);
    copyout_bytes(buf, host_buf, buf_len);
    _free(host_path);
    return ret;
},

path_rename: function(fd0, path0, path_len0, fd1, path1, path_len1) {
    let host_path0 = copyin_bytes(path0, path_len0);
    let host_path1 = copyin_bytes(path1, path_len1);
    let ret = ___wasi_path_rename(fd, host_path0, path_len0, fd1, host_path1, path1_len);
    _free(host_path1);
    _free(host_path0);
    return ret;
},

fd_filestat_get: function(fd, buf) {
    let host_buf = _malloc(56); // sizeof __wasi_filestat_t
    let ret = ___wasi_fd_filestat_get(host_buf);
    copyout_bytes(buf, host_buf, 56);
    return ret;
},

fd_filestat_set_size: function(fd, size) {
    return ___wasi_fd_filestat_set_size(fd, size);
},

fd_filestat_set_times: function(fd, st_atim, st_mtim, fstflags) {
    return ___wasi_fd_filestat_set_times(fd, st_atim, st_mtim, fstflags);
},

path_filestat_get: function(fd, path, path_len, buf) {
    let host_path = copyin_bytes(path, path_len);
    let host_buf = _malloc(56); // sizeof __wasi_filestat_t
    let ret = ___wasi_path_filestat_get(fd, host_path, path_len, host_buf);
    copyout_bytes(buf, host_buf, 56);
    _free(host_path);
    return ret;
},

path_filestat_set_times: function(fd, path, path_len, st_atim, st_mtim, flags) {
    let host_path = copyin_bytes(path, path_len);
    let ret = ___wasi_path_filestat_set_times(fd, host_path, st_atim, st_mtim, fstflags);
    _free(host_path);
    return ret;
},

path_symlink: function(path0, path_len0, fd, path1, path_len1) {
    let host_path0 = copyin_bytes(path0, path0_len);
    let host_path1 = copyin_bytes(path1, path1_len);
    let ret = ___wasi_path_symlink(host_path0, path_len0, fd, host_path1, path_len1);
    _free(host_path1);
    _free(host_path0);
    return ret;
},

path_unlink_file: function(fd, path, path_len, flags) {
    let host_path = copyin_bytes(path, path_len);
    let ret = ___wasi_path_unlink_file(fd, host_path, path_len, flags);
    _free(host_path);
    return ret;
},

path_remove_directory: function(fd, path, path_len, flags) {
    let host_path = copyin_bytes(path, path_len);
    let ret = ___wasi_path_remove_directory(fd, host_path, path_len, flags);
    _free(host_path);
    return ret;
},

poll_oneoff: function(in_, out, nsubscriptions, nevents) {
    let host_in = copyin_bytes(in_, nsubscriptions * 56); // sizeof __wasi_subscription_t
    let host_out = _malloc(nsubscriptions * 32); // sizeof __wasi_event_t
    let host_nevents = _malloc(4);
    let ret = ___wasi_poll_oneoff(host_in, host_out, host_nevents);
    copyout_bytes(out, host_out, nsubscriptions * 32);
    copyout_i32(nevents, host_nevents);
    _free(host_in);
    return ret;
},

proc_exit: function(rval) {
    let message;
    if (rval == 0) {
        message = "success";
    } else {
        message = "error code " + rval;
    }
    throw new WASIExit(rval, message);
},

proc_raise: function(sig) {
    if (sig == 18 || // SIGSTOP
        sig == 19 || // SIGTSTP
        sig == 20 || // SIGTTIN
        sig == 21 || // SIGTTOU
        sig == 22 || // SIGURG
        sig == 16 || // SIGCHLD
        sig == 13)   // SIGPIPE
    {
      return 0;
    }

    let message = "raised signal " + sig;
    throw new WASIExit(128 + sig, message);
},

random_get: function(buf, buf_len) {
    let host_buf = _malloc(buf_len);
    let ret = ___wasi_random_get(host_buf, buf_len);
    copyout_bytes(buf, host_buf, buf_len);
    return ret;
},

sched_yield: function() {
    return ___wasi_sched_yield();
},

sock_recv: function(sock, ri_data, ri_data_len, ri_flags, ro_datalen, ro_flags) {
    let host_ri_data = translate_iovs(ri_data, ri_data_len);
    let host_ro_datalen = _malloc(4);
    let ret = ___wasi_sock_recv(sock, host_ri_data, ri_data_len, ri_flags, host_ro_data, ro_flags);
    copyout_i32(ro_datalen, host_ro_datalen);
    free_iovs(host_ri_data, ri_data_len);
    return ret;
},

sock_send: function(sock, si_data, si_data_len, si_flags, so_datalen) {
    let host_si_data = translate_ciovs(si_data, si_data_len);
    let host_so_datalen = _malloc(4);
    let ret = ___wasi_sock_send(sock, host_si_data, si_data_len, si_flags, host_so_datalen);
    copyout_i32(so_datalen, host_so_datalen);
    free_ciovs(host_si_data, si_data_len);
    return ret;
},

sock_shutdown: function(sock, how) {
    return ___wasi_sock_shutdown(sock, how);
}

};

File: crates/wiggle/tests/wasi.rs
use wiggle::{GuestErrorType, GuestMemory, GuestPtr};
use wiggle_test::WasiCtx;

// This test file exists to make sure that the entire `wasi.witx` file can be
// handled by wiggle, producing code that compiles correctly.
// The trait impls here are never executed, and just exist to validate that the
// witx is exposed with the type signatures that we expect.

wiggle::from_witx!({
    witx: ["$CARGO_MANIFEST_DIR/tests/wasi.witx"],
});

// The only test in this file is to verify that the witx document provided by the
// proc macro in the `metadata` module is equal to the document on the disk.
#[test]
fn document_equivalent() {
    let macro_doc = metadata::document();
    let mut path = std::path::PathBuf::from(env!("CARGO_MANIFEST_DIR"));
    path.push("tests");
    path.push("wasi.witx");
    let disk_doc = witx::load(&[path]).expect("load wasi.witx from disk");

    assert_eq!(macro_doc, disk_doc);
}

type Result<T> = std::result::Result<T, types::Errno>;

impl GuestErrorType for types::Errno {
    fn success() -> types::Errno {
        types::Errno::Success
    }
}

impl<'a> crate::wasi_snapshot_preview1::WasiSnapshotPreview1 for WasiCtx<'a> {
    fn args_get(
        &mut self,
        _memory: &mut GuestMemory<'_>,
        _argv: GuestPtr<GuestPtr<u8>>,
        _argv_buf: GuestPtr<u8>,
    ) -> Result<()> {
        unimplemented!("args_get")
    }

    fn args_sizes_get(
        &mut self,
        _memory: &mut GuestMemory<'_>,
    ) -> Result<(types::Size, types::Size)> {
        unimplemented!("args_sizes_get")
    }

    fn environ_get(
        &mut self,
        _memory: &mut GuestMemory<'_>,
        _environ: GuestPtr<GuestPtr<u8>>,
        _environ_buf: GuestPtr<u8>,
    ) -> Result<()> {
        unimplemented!("environ_get")
    }

    fn environ_sizes_get(
        &mut self,
        _memory: &mut GuestMemory<'_>,
    ) -> Result<(types::Size, types::Size)> {
        unimplemented!("environ_sizes_get")
    }

    fn clock_res_get(
        &mut self,
        _memory: &mut GuestMemory<'_>,
        _id: types::Clockid,
    ) -> Result<types::Timestamp> {
        unimplemented!("clock_res_get")
    }

    fn clock_time_get(
        &mut self,
        _memory: &mut GuestMemory<'_>,
        _id: types::Clockid,
        _precision: types::Timestamp,
    ) -> Result<types::Timestamp> {
        unimplemented!("clock_time_get")
    }

    fn fd_advise(
        &mut self,
        _memory: &mut GuestMemory<'_>,
        _fd: types::Fd,
        _offset: types::Filesize,
        _len: types::Filesize,
        _advice: types::Advice,
    ) -> Result<()> {
        unimplemented!("fd_advise")
    }

    fn fd_allocate(
        &mut self,
        _memory: &mut GuestMemory<'_>,
        _fd: types::Fd,
        _offset: types::Filesize,
        _len: types::Filesize,
    ) -> Result<()> {
        unimplemented!("fd_allocate")
    }

    fn fd_close(&mut self, _memory: &mut GuestMemory<'_>, _fd: types::Fd) -> Result<()> {
        unimplemented!("fd_close")
    }

    fn fd_datasync(&mut self, _memory: &mut GuestMemory<'_>, _fd: types::Fd) -> Result<()> {
        unimplemented!("fd_datasync")
    }

    fn fd_fdstat_get(
        &mut self,
        _memory: &mut GuestMemory<'_>,
        _fd: types::Fd,
    ) -> Result<types::Fdstat> {
        unimplemented!("fd_fdstat_get")
    }

    fn fd_fdstat_set_flags(
        &mut self,
        _memory: &mut GuestMemory<'_>,
        _fd: types::Fd,
        _flags: types::Fdflags,
    ) -> Result<()> {
        unimplemented!("fd_fdstat_set_flags")
    }

    fn fd_fdstat_set_rights(
        &mut self,
        _memory: &mut GuestMemory<'_>,
        _fd: types::Fd,
        _fs_rights_base: types::Rights,
        _fs_rights_inherting: types::Rights,
    ) -> Result<()> {
        unimplemented!("fd_fdstat_set_rights")
    }

    fn fd_filestat_get(
        &mut self,
        _memory: &mut GuestMemory<'_>,
        _fd: types::Fd,
    ) -> Result<types::Filestat> {
        unimplemented!("fd_filestat_get")
    }

    fn fd_filestat_set_size(
        &mut self,
        _memory: &mut GuestMemory<'_>,
        _fd: types::Fd,
        _size: types::Filesize,
    ) -> Result<()> {
        unimplemented!("fd_filestat_set_size")
    }

    fn fd_filestat_set_times(
        &mut self,
        _memory: &mut GuestMemory<'_>,
        _fd: types::Fd,
        _atim: types::Timestamp,
        _mtim: types::Timestamp,
        _fst_flags: types::Fstflags,
    ) -> Result<()> {
        unimplemented!("fd_filestat_set_times")
    }

    fn fd_pread(
        &mut self,
        _memory: &mut GuestMemory<'_>,
        _fd: types::Fd,
        _iovs: types::IovecArray,
        _offset: types::Filesize,
    ) -> Result<types::Size> {
        unimplemented!("fd_pread")
    }

    fn fd_prestat_get(
        &mut self,
        _memory: &mut GuestMemory<'_>,
        _fd: types::Fd,
    ) -> Result<types::Prestat> {
        unimplemented!("fd_prestat_get")
    }

    fn fd_prestat_dir_name(
        &mut self,
        _memory: &mut GuestMemory<'_>,
        _fd: types::Fd,
        _path: GuestPtr<u8>,
        _path_len: types::Size,
    ) -> Result<()> {
        unimplemented!("fd_prestat_dir_name")
    }

    fn fd_pwrite(
        &mut self,
        _memory: &mut GuestMemory<'_>,
        _fd: types::Fd,
        _ciovs: types::CiovecArray,
        _offset: types::Filesize,
    ) -> Result<types::Size> {
        unimplemented!("fd_pwrite")
    }

    fn fd_read(
        &mut self,
        _memory: &mut GuestMemory<'_>,
        _fd: types::Fd,
        _iovs: types::IovecArray,
    ) -> Result<types::Size> {
        unimplemented!("fd_read")
    }

    fn fd_readdir(
        &mut self,
        _memory: &mut GuestMemory<'_>,
        _fd: types::Fd,
        _buf: GuestPtr<u8>,
        _buf_len: types::Size,
        _cookie: types::Dircookie,
    ) -> Result<types::Size> {
        unimplemented!("fd_readdir")
    }

    fn fd_renumber(
        &mut self,
        _memory: &mut GuestMemory<'_>,
        _fd: types::Fd,
        _to: types::Fd,
    ) -> Result<()> {
        unimplemented!("fd_renumber")
    }

    fn fd_seek(
        &mut self,
        _memory: &mut GuestMemory<'_>,
        _fd: types::Fd,
        _offset: types::Filedelta,
        _whence: types::Whence,
    ) -> Result<types::Filesize> {
        unimplemented!("fd_seek")
    }

    fn fd_sync(&mut self, _memory: &mut GuestMemory<'_>, _fd: types::Fd) -> Result<()> {
        unimplemented!("fd_sync")
    }

    fn fd_tell(
        &mut self,
        _memory: &mut GuestMemory<'_>,
        _fd: types::Fd,
    ) -> Result<types::Filesize> {
        unimplemented!("fd_tell")
    }

    fn fd_write(
        &mut self,
        _memory: &mut GuestMemory<'_>,
        _fd: types::Fd,
        _ciovs: types::CiovecArray,
    ) -> Result<types::Size> {
        unimplemented!("fd_write")
    }

    fn path_create_directory(
        &mut self,
        _memory: &mut GuestMemory<'_>,
        _fd: types::Fd,
        _path: GuestPtr<str>,
    ) -> Result<()> {
        unimplemented!("path_create_directory")
    }

    fn path_filestat_get(
        &mut self,
        _memory: &mut GuestMemory<'_>,
        _fd: types::Fd,
        _flags: types::Lookupflags,
        _path: GuestPtr<str>,
    ) -> Result<types::Filestat> {
        unimplemented!("path_filestat_get")
    }

    fn path_filestat_set_times(
        &mut self,
        _memory: &mut GuestMemory<'_>,
        _fd: types::Fd,
        _flags: types::Lookupflags,
        _path: GuestPtr<str>,
        _atim: types::Timestamp,
        _mtim: types::Timestamp,
        _fst_flags: types::Fstflags,
    ) -> Result<()> {
        unimplemented!("path_filestat_set_times")
    }

    fn path_link(
        &mut self,
        _memory: &mut GuestMemory<'_>,
        _old_fd: types::Fd,
        _old_flags: types::Lookupflags,
        _old_path: GuestPtr<str>,
        _new_fd: types::Fd,
        _new_path: GuestPtr<str>,
    ) -> Result<()> {
        unimplemented!("path_link")
    }

    fn path_open(
        &mut self,
        _memory: &mut GuestMemory<'_>,
        _fd: types::Fd,
        _dirflags: types::Lookupflags,
        _path: GuestPtr<str>,
        _oflags: types::Oflags,
        _fs_rights_base: types::Rights,
        _fs_rights_inherting: types::Rights,
        _fdflags: types::Fdflags,
    ) -> Result<types::Fd> {
        unimplemented!("path_open")
    }

    fn path_readlink(
        &mut self,
        _memory: &mut GuestMemory<'_>,
        _fd: types::Fd,
        _path: GuestPtr<str>,
        _buf: GuestPtr<u8>,
        _buf_len: types::Size,
    ) -> Result<types::Size> {
        unimplemented!("path_readlink")
    }

    fn path_remove_directory(
        &mut self,
        _memory: &mut GuestMemory<'_>,
        _fd: types::Fd,
        _path: GuestPtr<str>,
    ) -> Result<()> {
        unimplemented!("path_remove_directory")
    }

    fn path_rename(
        &mut self,
        _memory: &mut GuestMemory<'_>,
        _fd: types::Fd,
        _old_path: GuestPtr<str>,
        _new_fd: types::Fd,
        _new_path: GuestPtr<str>,
    ) -> Result<()> {
        unimplemented!("path_rename")
    }

    fn path_symlink(
        &mut self,
        _memory: &mut GuestMemory<'_>,
        _old_path: GuestPtr<str>,
        _fd: types::Fd,
        _new_path: GuestPtr<str>,
    ) -> Result<()> {
        unimplemented!("path_symlink")
    }

    fn path_unlink_file(
        &mut self,
        _memory: &mut GuestMemory<'_>,
        _fd: types::Fd,
        _path: GuestPtr<str>,
    ) -> Result<()> {
        unimplemented!("path_unlink_file")
    }

    fn poll_oneoff(
        &mut self,
        _memory: &mut GuestMemory<'_>,
        _in_: GuestPtr<types::Subscription>,
        _out: GuestPtr<types::Event>,
        _nsubscriptions: types::Size,
    ) -> Result<types::Size> {
        unimplemented!("poll_oneoff")
    }

    fn proc_exit(
        &mut self,
        _memory: &mut GuestMemory<'_>,
        _rval: types::Exitcode,
    ) -> anyhow::Error {
        unimplemented!("proc_exit")
    }

    fn proc_raise(&mut self, _memory: &mut GuestMemory<'_>, _sig: types::Signal) -> Result<()> {
        unimplemented!("proc_raise")
    }

    fn sched_yield(&mut self, _memory: &mut GuestMemory<'_>) -> Result<()> {
        unimplemented!("sched_yield")
    }

    fn random_get(
        &mut self,
        _memory: &mut GuestMemory<'_>,
        _buf: GuestPtr<u8>,
        _buf_len: types::Size,
    ) -> Result<()> {
        unimplemented!("random_get")
    }

    fn sock_recv(
        &mut self,
        _memory: &mut GuestMemory<'_>,
        _fd: types::Fd,
        _ri_data: types::IovecArray,
        _ri_flags: types::Riflags,
    ) -> Result<(types::Size, types::Roflags)> {
        unimplemented!("sock_recv")
    }

    fn sock_send(
        &mut self,
        _memory: &mut GuestMemory<'_>,
        _fd: types::Fd,
        _si_data: types::CiovecArray,
        _si_flags: types::Siflags,
    ) -> Result<types::Size> {
        unimplemented!("sock_send")
    }

    fn sock_shutdown(
        &mut self,
        _memory: &mut GuestMemory<'_>,
        _fd: types::Fd,
        _how: types::Sdflags,
    ) -> Result<()> {
        unimplemented!("sock_shutdown")
    }
}
